File: Calculator/SetCalculator.java
package Calculator;
import java.util.ArrayList;

import Calculator.Attributes.Color;
import Calculator.Attributes.Num;
import Calculator.Attributes.Shading;
import Calculator.Attributes.Shape;

public class SetCalculator {
    private Card[] cards;
    private ArrayList<Card[]> sets;
    public SetCalculator(Card ... cards){
        this.cards = cards;
        if(cards.length < 3 || cards.length % 3 != 0){
            throw new IllegalArgumentException("Invalid Card Amount");
        }
        sets = new ArrayList<>();
        findAllSets();
    }

    public int getSize(){
        return cards.length;
    }
    
    private static boolean checkNum(Card a, Card b, Card c){
        if(a.getNumber().ordinal() != b.getNumber().ordinal() &&
           b.getNumber().ordinal() != c.getNumber().ordinal() &&
           c.getNumber().ordinal() != a.getNumber().ordinal() ){
            return true;
        }
        if(a.getNumber().ordinal() == b.getNumber().ordinal() && b.getNumber().ordinal() == c.getNumber().ordinal()){
            return true;
        }
        return false;
    }

    private static boolean checkColor(Card a, Card b, Card c){
        if(a.getColor().ordinal() != b.getColor().ordinal() &&
           b.getColor().ordinal() != c.getColor().ordinal() &&
           c.getColor().ordinal() != a.getColor().ordinal() ){
            return true;
        }
        if(a.getColor().ordinal() == b.getColor().ordinal() && b.getColor().ordinal() == c.getColor().ordinal()){
            return true;
        }
        return false;
        
    }
     
    private static boolean checkShape(Card a, Card b, Card c){
        if(a.getShape().ordinal() != b.getShape().ordinal() &&
           b.getShape().ordinal() != c.getShape().ordinal() &&
           c.getShape().ordinal() != a.getShape().ordinal() ){
            return true;
        }
        if(a.getShape().ordinal() == b.getShape().ordinal() && b.getShape().ordinal() == c.getShape().ordinal()){
            return true;
        }
        return false;
    }
    
    private static boolean checkShading(Card a, Card b, Card c){
        if(a.getShading().ordinal() != b.getShading().ordinal() &&
           b.getShading().ordinal() != c.getShading().ordinal() &&
           c.getShading().ordinal() != a.getShading().ordinal() ){
            return true;
        }
        if(a.getShading().ordinal() == b.getShading().ordinal() && b.getShading().ordinal() == c.getShading().ordinal()){
            return true;
        }
        return false;
    }

    private static boolean isSet(Card a, Card b, Card c){
        return checkColor(a, b, c) && checkNum(a, b, c) && checkShading(a, b, c) && checkShape(a, b, c);
    }
    
    

    public void findAllSets(){
        
        for(int i = 0; i < cards.length - 2; i++){
            Card firstCard = cards[i];
            
            for(int j = i + 1; j < cards.length - 1; j++){
                Card secondCard = cards[j];
                
                for(int k = j + 1; k < cards.length; k++){
                    Card thirdCard = cards[k];
                    if(isSet(firstCard, secondCard, thirdCard)){
                        sets.add(new Card[] { firstCard, secondCard, thirdCard});
                    }
                }
            }
        }
    }

    public String toString(){
        StringBuilder result = new StringBuilder("Sets:\n");

        for (Card[] set : sets) {
            result.append("[ ");
            for (Card card : set) {
                result.append(card).append(", ");
            }
            result.delete(result.length() - 2, result.length()); // Remove the trailing ", "
            result.append(" ]\n");
        }

        return result.toString();
    }
    
    public static void main(String[] args){
        Card[] cards = new Card[] {
            new Card(Num.TWO, Color.PURPLE, Shading.SOLID, Shape.OVAL),
            new Card(Num.ONE, Color.RED, Shading.STRIPE, Shape.OVAL),
            new Card(Num.THREE,Color.RED, Shading.STRIPE, Shape.SQUIGGLE),
            
            new Card(Num.ONE, Color.PURPLE, Shading.SOLID, Shape.DIAMOND),
            new Card(Num.THREE, Color.PURPLE, Shading.EMPTY, Shape.DIAMOND),
            new Card(Num.TWO, Color.GREEN, Shading.STRIPE, Shape.DIAMOND),
            
            new Card(Num.ONE, Color.GREEN, Shading.STRIPE, Shape.DIAMOND),
            new Card(Num.ONE, Color.GREEN, Shading.SOLID, Shape.DIAMOND),
            new Card(Num.TWO, Color.GREEN, Shading.SOLID, Shape.DIAMOND),

            new Card(Num.THREE, Color.RED, Shading.SOLID, Shape.DIAMOND),
            new Card(Num.ONE, Color.GREEN, Shading.STRIPE, Shape.OVAL),
            new Card(Num.ONE, Color.GREEN, Shading.EMPTY, Shape.OVAL),

        };

        SetCalculator calculator = new SetCalculator(cards);
        System.out.println(calculator);

    }


}

File: Python Calc/set_calculator.py
from enum import Enum, auto
from itertools import combinations

# Attribues
class color(Enum):
    RED = 1
    GREEN = 2
    PURPLE = 3

class num(Enum):
    ONE = 1
    TWO = 2
    THREE = 3

class shading(Enum):
    SOLID = 1
    STRIPE = 2
    EMPTY = 3

class shape(Enum):
    DIAMOND = 1
    SQUIGGLE = 2
    OVAL = 3

class card:
    def __init__(self, number, color, shading, shape):
        self.number = number
        self.color = color
        self.shading = shading
        self.shape = shape

    def __str__(self):
        return f"({self.number.name}, {self.color.name}, {self.shading.name}, {self.shape.name})"
    
class set_calculator: 
    def __init__(self, *cards):
        self.cards = cards
        if len(cards) < 3 or len(cards) % 3 != 0:
            raise ValueError("Invalid Card Amount")
        self.sets = []
        self.find_all_sets()

    def check_num( a, b, c):
        return len(set([a.number, b.number, c.number])) != 2
    
    def check_color( a, b, c):
        return len(set([a.color, b.color, c.color])) != 2
    
    def check_shading(a, b, c):
        return len(set([a.shading, b.shading, c.shading])) != 2
    
    def check_shape( a, b, c):
        return len(set([a.shape, b.shape, c.shape])) != 2
    
    def is_set(self, a, b, c):
        return all([set_calculator.check_num(a, b, c), 
                   set_calculator.check_color(a, b, c),
                   set_calculator.check_shading(a, b, c), 
                   set_calculator.check_shape(a, b, c)])
    def find_all_sets(self):
        for i in range(len(self.cards) - 2):
            first_card = self.cards[i]
            for j in range(i + 1, len(self.cards) - 1):
                second_card = self.cards[j]
                for k in range(j + 1, len(self.cards)):
                    third_card = self.cards[k]

                    if self.is_set(first_card, second_card, third_card):
                        self.sets.append((first_card, second_card, third_card))

    def __str__(self):
        result = "Sets:\n"

        for s in self.sets:
            result += "[ "
            for card in s:
                result += str(card) + ", "
            result = result[:-2] + " ]\n"

        return result
            
    
cards = (
    # Set 1
    card(num.TWO, color.PURPLE, shading.SOLID, shape.OVAL),
    card(num.TWO, color.GREEN, shading.STRIPE, shape.OVAL),
    card(num.TWO, color.RED, shading.EMPTY, shape.OVAL),

    # Set 2
    card(num.ONE, color.RED, shading.STRIPE, shape.OVAL),
    card(num.ONE, color.PURPLE, shading.SOLID, shape.OVAL),
    card(num.ONE, color.GREEN, shading.EMPTY, shape.OVAL),

    # Set 3
    card(num.THREE, color.PURPLE, shading.EMPTY, shape.DIAMOND),
    card(num.THREE, color.GREEN, shading.STRIPE, shape.DIAMOND),
    card(num.THREE, color.RED, shading.SOLID, shape.DIAMOND),

    # Set 4
    card(num.TWO, color.GREEN, shading.SOLID, shape.DIAMOND),
    card(num.THREE, color.PURPLE, shading.SOLID, shape.DIAMOND),
    card(num.ONE, color.RED, shading.STRIPE, shape.DIAMOND),

    # Set 5
    card(num.THREE, color.PURPLE, shading.SOLID, shape.SQUIGGLE),
    card(num.TWO, color.GREEN, shading.STRIPE, shape.SQUIGGLE),
    card(num.ONE, color.RED, shading.EMPTY, shape.SQUIGGLE),
)

calculator = set_calculator(*cards)
print(calculator)

